src/msg/simple/Accepter.cc

```diff
@@ -263,50 +321,84 @@ void *Accepter::entry()
     if (sd >= 0) {
       int r = set_close_on_exec(sd);
       if (r) {
-	ldout(msgr->cct,0) << "accepter set_close_on_exec() failed "
+	ldout(msgr->cct,0) << __func__ << " set_close_on_exec() failed "
 	      << cpp_strerror(r) << dendl;
       }
       errors = 0;
-      ldout(msgr->cct,10) << "accepted incoming on sd " << sd << dendl;
+      ldout(msgr->cct,10) << __func__ << " incoming on sd " << sd << dendl;
       
       msgr->add_accept_pipe(sd);
     } else {
-      ldout(msgr->cct,0) << "accepter no incoming connection?  sd = " << sd
+      ldout(msgr->cct,0) << __func__ << " no incoming connection?  sd = " << sd
 	      << " errno " << errno << " " << cpp_strerror(errno) << dendl;
       if (++errors > 4)
 	break;
     }
   }
 
-  ldout(msgr->cct,20) << "accepter closing" << dendl;
+  ldout(msgr->cct,20) << __func__ << " closing" << dendl;
   // don't close socket, in case we start up again?  blech.
-  if (listen_sd >= 0) {
-    ::close(listen_sd);
-    listen_sd = -1;
-  }
-  ldout(msgr->cct,10) << "accepter stopping" << dendl;
+  // if (listen_sd >= 0) {
+  //   ::close(listen_sd);
+  //   listen_sd = -1;
+  // }
+  if (shutdown_rd_fd >= 0) {
+       ::close(shutdown_rd_fd);
+       shutdown_rd_fd = -1;
+   }
+
+  ldout(msgr->cct,10) << __func__ << " stopping" << dendl;
   return 0;
 }
 
 void Accepter::stop()
 {
   done = true;
-  ldout(msgr->cct,10) << "stop accepter" << dendl;
-
-  if (listen_sd >= 0) {
-    ::shutdown(listen_sd, SHUT_RDWR);
+  ldout(msgr->cct,10) << __func__ << " accept listening on: " << listen_sd << dendl;
+
+  if (shutdown_wr_fd < 0)
+    return;
+
+  // Send a byte to the shutdown pipe that the thread is listening to
+  char buf[1] = { 0x0 };
+  int ret = safe_write(shutdown_wr_fd, buf, 1);
+  if (ret < 0) {
+      ldout(msgr->cct,10) << __func__ << "close failed: "
+             << " errno " << errno << " " << cpp_strerror(errno) << dendl;
+  }  else {
+      ldout(msgr->cct,15) << __func__ << " signaled poll" << dendl;
   }
+  VOID_TEMP_FAILURE_RETRY(close(shutdown_wr_fd));
+  shutdown_wr_fd = -1;
 
   // wait for thread to stop before closing the socket, to avoid
   // racing against fd re-use.
   if (is_started()) {
+      ldout(msgr->cct,0) << __func__ << " wait for thread to join." << dendl;
     join();
   }
 
   if (listen_sd >= 0) {
-    ::close(listen_sd);
+    if (::close(listen_sd) < 0) {
+      ldout(msgr->cct,10) << __func__ << "close listen_sd failed: "
+	      << " errno " << errno << " " << cpp_strerror(errno) << dendl;
+    }
     listen_sd = -1;
   }
+  if (shutdown_rd_fd >= 0) {
+	if (::close(shutdown_rd_fd) < 0) {
+      ldout(msgr->cct,10) << __func__ << "close shutdown_rd_fd failed: "
+	      << " errno " << errno << " " << cpp_strerror(errno) << dendl;
+    }
+	shutdown_rd_fd = -1;
+  }
+  if (shutdown_wr_fd >= 0) {
```

Comment:
@yuyuyu101 
You mean, it should have been already closed by the other part of the thread?
But it does not hurt to have it here?

It is here to really prevent fd from "escaping" while still open.
Perhaps I should first make this work with an assert, just to be safe and it really doesn't allow file-handle leakage.